File: nativelib/src/main/scala/scala/scalanative/posix/sys/socket.scala
package scala.scalanative
package posix
package sys

import scalanative.native._

@extern
object socket {
  type socklen_t   = CUnsignedInt
  type sa_family_t = CUnsignedShort
  type _14         = Nat.Digit[Nat._1, Nat._4]
  type sockaddr =
    CStruct2[sa_family_t, // sa_family
             CArray[CChar, _14]] // sa_data, size = 14 in OS X and Linux
  type sockaddr_storage = CStruct1[sa_family_t] // ss_family
  type msghdr = CStruct7[Ptr[Byte], // msg_name
                         socklen_t, // msg_namelen
                         Ptr[uio.iovec], // msg_iov
                         CInt, // msg_iovlen
                         Ptr[Byte], // msg_control
                         socklen_t, // msg_crontrollen
                         CInt] // msg_flags
  type cmsghdr = CStruct3[socklen_t, // cmsg_len
                          CInt, // cmsg_level
                          CInt] // cmsg_type
  type iovec = uio.iovec
  type linger = CStruct2[CInt, // l_onoff
                         CInt] // l_linger

  @name("scalanative_SCM_RIGHTS")
  def SCM_RIGHTS: CInt = extern

  @name("scalanative_SOCK_DGRAM")
  def SOCK_DGRAM: CInt = extern

  @name("scalanative_SOCK_RAW")
  def SOCK_RAW: CInt = extern

  @name("scalanative_SOCK_SEQPACKET")
  def SOCK_SEQPACKET: CInt = extern

  @name("scalanative_SOCK_STREAM")
  def SOCK_STREAM: CInt = extern

  @name("scalanative_SOL_SOCKET")
  def SOL_SOCKET: CInt = extern

  @name("scalanative_SO_ACCEPTCONN")
  def SO_ACCEPTCONN: CInt = extern

  @name("scalanative_SO_BROADCAST")
  def SO_BROADCAST: CInt = extern

  @name("scalanative_SO_DEBUG")
  def SO_DEBUG: CInt = extern

  @name("scalanative_SO_DONTROUTE")
  def SO_DONTROUTE: CInt = extern

  @name("scalanative_SO_ERROR")
  def SO_ERROR: CInt = extern

  @name("scalanative_SO_KEEPALIVE")
  def SO_KEEPALIVE: CInt = extern

  @name("scalanative_SO_LINGER")
  def SO_LINGER: CInt = extern

  @name("scalanative_SO_OOBINLINE")
  def SO_OOBINLINE: CInt = extern

  @name("scalanative_SO_RCVBUF")
  def SO_RCVBUF: CInt = extern

  @name("scalanative_SO_RCVLOWAT")
  def SO_RCVLOWAT: CInt = extern

  @name("scalanative_SO_RCVTIMEO")
  def SO_RCVTIMEO: CInt = extern

  @name("scalanative_SO_REUSEADDR")
  def SO_REUSEADDR: CInt = extern

  @name("scalanative_SO_SNDBUF")
  def SO_SNDBUF: CInt = extern

  @name("scalanative_SO_SNDLOWAT")
  def SO_SNDLOWAT: CInt = extern

  @name("scalanative_SO_SNDTIMEO")
  def SO_SNDTIMEO: CInt = extern

  @name("scalanative_SO_TYPE")
  def SO_TYPE: CInt = extern

  @name("scalanative_SOMAXCONN")
  def SOMAXCONN: CInt = extern

  @name("scalanative_MSG_CTRUNC")
  def MSG_CTRUNC: CInt = extern

  @name("scalanative_MSG_DONTROUTE")
  def MSG_DONTROUTE: CInt = extern

  @name("scalanative_MSG_EOR")
  def MSG_EOR: CInt = extern

  @name("scalanative_MSG_OOB")
  def MSG_OOB: CInt = extern

  // Surprisingly, this doesn't exist on MacOS
  // @name("scalanative_MSG_NOSIGNAL")
  // def MSG_NOSIGNAL: CInt = extern

  @name("scalanative_MSG_PEEK")
  def MSG_PEEK: CInt = extern

  @name("scalanative_MSG_TRUNC")
  def MSG_TRUNC: CInt = extern

  @name("scalanative_MSG_WAITALL")
  def MSG_WAITALL: CInt = extern

  @name("scalanative_AF_INET")
  def AF_INET: CInt = extern

  @name("scalanative_AF_INET6")
  def AF_INET6: CInt = extern

  @name("scalanative_AF_UNIX")
  def AF_UNIX: CInt = extern

  @name("scalanative_AF_UNSPEC")
  def AF_UNSPEC: CInt = extern

  @name("scalanative_socket")
  def socket(domain: CInt, tpe: CInt, protocol: CInt): CInt = extern

  @name("scalanative_connect")
  def connect(socket: CInt,
              address: Ptr[sockaddr],
              address_len: socklen_t): CInt = extern

  @name("scalanative_bind")
  def bind(socket: CInt,
           address: Ptr[sockaddr],
           address_len: socklen_t): CInt = extern

  @name("scalanative_listen")
  def listen(socket: CInt, backlog: CInt): CInt = extern

  @name("scalanative_accept")
  def accept(socket: CInt,
             address: Ptr[sockaddr],
             address_len: Ptr[socklen_t]): CInt = extern

  @name("scalanative_setsockopt")
  def setsockopt(socket: CInt,
                 level: CInt,
                 option_name: CInt,
                 options_value: Ptr[Byte],
                 option_len: socklen_t): CInt = extern

  @name("scalanative_recv")
  def recv(socket: CInt,
           buffer: Ptr[Byte],
           length: CSize,
           flags: CInt): CSSize = extern

  @name("scalanative_send")
  def send(socket: CInt,
           buffer: Ptr[Byte],
           length: CSize,
           flags: CInt): CSSize = extern
}

object socketOps {
  import socket._

  implicit class sockaddrOps(val ptr: Ptr[sockaddr]) extends AnyVal {
    def sa_family: sa_family_t      = !ptr._1
    def sa_data: CArray[CChar, _14] = !ptr._2

    def sa_family_=(v: sa_family_t): Unit      = !ptr._1 = v
    def sa_data_=(v: CArray[CChar, _14]): Unit = !ptr._2 = v
  }

  implicit class sockaddr_storageOps(val ptr: Ptr[sockaddr_storage])
      extends AnyVal {
    def ss_family: sa_family_t            = !ptr._1
    def ss_family_=(v: sa_family_t): Unit = !ptr._1 = v
  }

  implicit class msghdrOps(val ptr: Ptr[msghdr]) extends AnyVal {
    def msg_name: Ptr[Byte]       = !ptr._1
    def msg_namelen: socklen_t    = !ptr._2
    def msg_iov: Ptr[uio.iovec]   = !ptr._3
    def msg_iovlen: CInt          = !ptr._4
    def msg_control: Ptr[Byte]    = !ptr._5
    def msg_controllen: socklen_t = !ptr._6
    def msg_flags: CInt           = !ptr._7

    def msg_name_=(v: Ptr[Byte]): Unit       = !ptr._1 = v
    def msg_namelen_=(v: socklen_t): Unit    = !ptr._2 = v
    def msg_iov_=(v: Ptr[uio.iovec]): Unit   = !ptr._3 = v
    def msg_iovlen_=(v: CInt): Unit          = !ptr._4 = v
    def msg_control_=(v: Ptr[Byte]): Unit    = !ptr._5 = v
    def msg_controllen_=(v: socklen_t): Unit = !ptr._6 = v
    def msg_flags_=(v: CInt): Unit           = !ptr._7 = v
  }

  implicit class cmsghdrOps(val ptr: Ptr[cmsghdr]) extends AnyVal {
    def cmsg_len: socklen_t = !ptr._1
    def cmsg_level: CInt    = !ptr._2
    def cmsg_type: CInt     = !ptr._3

    def cmsg_len_=(v: socklen_t): Unit = !ptr._1 = v
    def cmsg_level_=(v: CInt): Unit    = !ptr._2 = v
    def cmsg_type_=(v: CInt): Unit     = !ptr._3 = v
  }

  implicit class lingerOps(val ptr: Ptr[linger]) extends AnyVal {
    def l_onoff: CInt  = !ptr._1
    def l_linger: CInt = !ptr._2

    def l_onoff_=(v: CInt): Unit  = !ptr._1 = v
    def l_linger_=(v: CInt): Unit = !ptr._2 = v
  }

}

File: posixlib/src/main/resources/scala-native/posix/sys/socket.c
#include <netinet/in.h>
#include "../netinet/in.h"
#include <sys/socket.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include "socket_conversions.h"
#include "socket.h"

int scalanative_scm_rights() { return SCM_RIGHTS; }

int scalanative_sock_dgram() { return SOCK_DGRAM; }

int scalanative_sock_raw() { return SOCK_RAW; }

int scalanative_sock_seqpacket() { return SOCK_SEQPACKET; }

int scalanative_sock_stream() { return SOCK_STREAM; }

int scalanative_sol_socket() { return SOL_SOCKET; }

int scalanative_so_acceptconn() { return SO_ACCEPTCONN; }

int scalanative_so_broadcast() { return SO_BROADCAST; }

int scalanative_so_debug() { return SO_DEBUG; }

int scalanative_so_dontroute() { return SO_DONTROUTE; }

int scalanative_so_error() { return SO_ERROR; }

int scalanative_so_keepalive() { return SO_KEEPALIVE; }

int scalanative_so_linger() { return SO_LINGER; }

int scalanative_so_oobinline() { return SO_OOBINLINE; }

int scalanative_so_rcvbuf() { return SO_RCVBUF; }

int scalanative_so_rcvlowat() { return SO_RCVLOWAT; }

int scalanative_so_rcvtimeo() { return SO_RCVTIMEO; }

int scalanative_so_reuseaddr() { return SO_REUSEADDR; }

int scalanative_so_sndbuf() { return SO_SNDBUF; }

int scalanative_so_sndlowat() { return SO_SNDLOWAT; }

int scalanative_so_sndtimeo() { return SO_SNDTIMEO; }

int scalanative_so_type() { return SO_TYPE; }

int scalanative_somaxconn() { return SOMAXCONN; }

int scalanative_msg_ctrunc() { return MSG_CTRUNC; }

int scalanative_msg_dontroute() { return MSG_DONTROUTE; }

int scalanative_msg_eor() { return MSG_EOR; }

int scalanative_msg_oob() { return MSG_OOB; }

int scalanative_msg_nosignal() {
#ifdef MSG_NOSIGNAL
    return MSG_NOSIGNAL;
#endif
#ifndef MSG_NOSIGNAL
    return 0;
#endif
}

int scalanative_msg_peek() { return MSG_PEEK; }

int scalanative_msg_trunc() { return MSG_TRUNC; }

int scalanative_msg_waitall() { return MSG_WAITALL; }

int scalanative_af_inet() { return AF_INET; }

int scalanative_af_inet6() { return AF_INET6; }

int scalanative_af_unix() { return AF_UNIX; }

int scalanative_af_unspec() { return AF_UNSPEC; }

int scalanative_getsockname(int socket, struct scalanative_sockaddr *address,
                            socklen_t *address_len) {
    struct sockaddr *converted_address;
    int convert_result =
        scalanative_convert_sockaddr(address, &converted_address, address_len);

    int result;

    if (convert_result == 0) {
        result = getsockname(socket, converted_address, address_len);
        convert_result = scalanative_convert_scalanative_sockaddr(
            converted_address, address, address_len);

        if (convert_result != 0) {
            errno = convert_result;
            result = -1;
        }
    } else {
        errno = convert_result;
        result = -1;
    }

    free(converted_address);
    return result;
}

int scalanative_socket(int domain, int type, int protocol) {
    return socket(domain, type, protocol);
}

int scalanative_bind(int socket, struct scalanative_sockaddr *address,
                     socklen_t address_len) {
    struct sockaddr *converted_address;
    int convert_result =
        scalanative_convert_sockaddr(address, &converted_address, &address_len);

    int result;

    if (convert_result == 0) {
        result = bind(socket, converted_address, address_len);
    } else {
        errno = convert_result;
        result = -1;
    }

    free(converted_address);
    return result;
}

int scalanative_connect(int socket, struct scalanative_sockaddr *address,
                        socklen_t address_len) {
    struct sockaddr *converted_address;
    int convert_result =
        scalanative_convert_sockaddr(address, &converted_address, &address_len);

    int result;

    if (convert_result == 0) {
        result = connect(socket, converted_address, address_len);
    } else {
        errno = convert_result;
        result = -1;
    }
    free(converted_address);
    return result;
}

int scalanative_listen(int socket, int backlog) {
    return listen(socket, backlog);
}

int scalanative_accept(int socket, struct scalanative_sockaddr *address,
                       socklen_t *address_len) {
    struct sockaddr *converted_address;
    int convert_result =
        scalanative_convert_sockaddr(address, &converted_address, address_len);

    int result;

    if (convert_result == 0) {
        result = accept(socket, converted_address, address_len);
        convert_result = scalanative_convert_scalanative_sockaddr(
            converted_address, address, address_len);

        if (convert_result != 0) {
            errno = convert_result;
            result = -1;
        }
    } else {
        errno = convert_result;
        result = -1;
    }

    free(converted_address);
    return result;
}

int scalanative_setsockopt(int socket, int level, int option_name,
                           void *option_value, socklen_t option_len) {
    return setsockopt(socket, level, option_name, option_value, option_len);
}

int scalanative_getsockopt(int socket, int level, int option_name,
                           void *option_value, socklen_t *option_len) {
    return getsockopt(socket, level, option_name, option_value, option_len);
}

int scalanative_recv(int socket, void *buffer, size_t length, int flags) {
    return recv(socket, buffer, length, flags);
}

int scalanative_send(int socket, void *buffer, size_t length, int flags) {
    return send(socket, buffer, length, flags);
}

int scalanative_shutdown(int socket, int how) { return shutdown(socket, how); }
